br/pkg/lightning/backend/external/iter.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"bytes"
+	"container/heap"
+	"context"
+	"io"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type heapElem interface {
+	sortKey() []byte
+}
+
+type sortedReader[T heapElem] interface {
+	path() string
+	next() (T, error)
+	close() error
+}
+
+type mergeHeapElem[T heapElem] struct {
+	elem      T
+	readerIdx int
+}
+
+type mergeHeap[T heapElem] []mergeHeapElem[T]
+
+func (h mergeHeap[T]) Len() int {
+	return len(h)
+}
+
+func (h mergeHeap[T]) Less(i, j int) bool {
+	return bytes.Compare(h[i].elem.sortKey(), h[j].elem.sortKey()) < 0
+}
+
+func (h mergeHeap[T]) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *mergeHeap[T]) Push(x interface{}) {
+	*h = append(*h, x.(mergeHeapElem[T]))
+}
+
+func (h *mergeHeap[T]) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+type mergeIter[T heapElem, R sortedReader[T]] struct {
+	h             mergeHeap[T]
+	readers       []*R
+	curr          T
```

Comment:
When `T` equals to `kvPair`, this may introduce unnecessary memory copy for each element, is this acceptable?

Reply:
Do you mean we should store *T?

Reply:
I made a simple bench https://github.com/pingcap/tidb/pull/45913/commits/8e1b184fc3c53bab340cf4a994cbb78c9cf1dec6

```
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/br/pkg/lightning/backend/external
cpu: Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz
BenchmarkValueT
BenchmarkValueT-12      	      14	  94879397 ns/op
BenchmarkPointerT
BenchmarkPointerT-12    	      25	  97058542 ns/op
PASS
```

value type `T` is slightly more quickly. Please help to check the bench mode

Reply:
I think 2.3% is fine, let's keep this.